ddtrace/contrib/graphql/patch.py

```diff
@@ -0,0 +1,189 @@
+from graphql.execution import ExecutionResult
+from graphql.language.source import Source
+
+from ddtrace import Span
+from ddtrace import config
+from ddtrace.constants import ANALYTICS_SAMPLE_RATE_KEY
+from ddtrace.constants import SPAN_MEASURED_KEY
+from ddtrace.internal.compat import stringify
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.internal.utils.version import parse_version
+from ddtrace.pin import Pin
+from ddtrace.vendor.wrapt import wrap_function_wrapper as _w
+
+from .. import trace_utils
+from ...ext import SpanTypes
+
+
+config._add("graphql", dict(_default_service="graphql"))
+
+
+graphql_version_str = getattr(graphql, "__version__")
+graphql_version = parse_version(graphql_version_str)
+
+if graphql_version < (3, 0):
+    from graphql.language.ast import Document
+else:
+    from graphql.language.ast import DocumentNode as Document
+
+
+def patch():
+    if getattr(graphql, "_datadog_patch", False) or graphql_version < (2, 0):
+        return
+
+    setattr(graphql, "_datadog_patch", True)
+
+    # patch graphql.graphql() and graphql.graphql_sync()
+    if graphql_version < (3, 0):
+        _w("graphql.graphql", "execute_graphql", _traced_operation("graphql.query"))
+    else:
+        _w("graphql.graphql", "graphql_impl", _traced_operation("graphql.query"))
+
+    # patch parse() used in graphql.graphql and graphql.graphql_sync
+    if graphql_version < (2, 1):
+        _w("graphql.graphql", "parse", _traced_operation("graphql.parse"))
+    elif graphql_version < (3, 0):
+        _w("graphql.backend.core", "parse", _traced_operation("graphql.parse"))
+    else:
+        _w("graphql.graphql", "parse", _traced_operation("graphql.parse"))
+
+    # patch validate() used in graphql.graphql and graphql.graphql_sync
+    if graphql_version < (2, 1):
+        _w("graphql.graphql", "validate", _traced_operation("graphql.validate"))
+    elif graphql_version < (3, 0):
+        _w("graphql.backend.core", "validate", _traced_operation("graphql.validate"))
+    else:
+        _w("graphql.validation", "validate", _traced_operation("graphql.validate"))
+
+    # patch execute() function used in graphql.graphql and graphql.graphql_sync
+    if (2, 1) <= graphql_version < (3, 0):
+        _w("graphql.backend.core", "execute", _traced_operation("graphql.execute"))
+    else:
+        _w("graphql.graphql", "execute", _traced_operation("graphql.execute"))
+
+    # Patch execute functions exposed in the public api
+    _w("graphql", "execute", _traced_operation("graphql.execute"))
+    _w("graphql.execution", "execute", _traced_operation("graphql.execute"))
+    if graphql_version < (3, 0):
+        _w("graphql.execution.executor", "execute", _traced_operation("graphql.execute"))
+    elif graphql_version >= (3, 1):
+        _w("graphql", "execute_sync", _traced_operation("graphql.execute"))
+        _w("graphql.execution", "execute_sync", _traced_operation("graphql.execute"))
+
+    # patch all resolvers
```

Comment:
Instead of monkey patching `resolve_field` I could instead pass a trace [middleware](https://docs.graphene-python.org/en/latest/execution/middleware/) to graphql.execute(). This middleware would wrap all resolvers in a span when resolve_field is called. Both approaches produce equivalent instrumentation (although using a trace middleware could provide more accurate span durations). I think the monkey patching approach is easier to follow.

ex: 
```
import graphql

class TracedResolverMiddleware(object):
    def resolve(self, next, root, info, **args):
        pin = Pin.get_from(graphql)
        if not pin or not pin.enabled():
            return next(root, info, **args)

        with pin.tracer.trace(
            name="graphql.resolve", 
            resource=info.field_name,
            service=trace_utils.int_service(pin, config.graphql),
            span_type=SpanTypes.WEB,
        ) as span:
            _init_span(span)
            return next(root, info, **args)

# wraps all resolvers in a span
result = graphql.execute(schema, source, middleware=[TracedResolverMiddleware])
```